packages/remix/src/ssr/utils.ts

```diff
@@ -0,0 +1,78 @@
+import cookie from 'cookie';
+
+import { AuthData } from './getAuthData';
+import { LoaderFunctionArgs, LoaderFunctionArgsWithAuth } from './types';
+
+/**
+ * Wraps obscured clerk internals with a readable `clerkState` key.
+ * This is intended to be passed by the user into <ClerkProvider>
+ *
+ * @internal
+ */
+export const wrapClerkState = (data: any) => {
+  return { clerkState: { __internal_clerk_state: { ...data } } };
+};
+
+/**
+ * Extracts the content inside the <html> tags of the interstitial page
+ *
+ * @internal
+ */
+export const extractInterstitialHtmlContent = (interstitial?: string) => {
+  try {
+    return (interstitial || '').match(/<html>([\s\S.]*)<\/html>/)![1];
+  } catch (e) {
+    throw new Error('This is a Clerk bug; Please report');
+  }
+};
+
+/**
+ * Inject the `auth` attribute to the SSR provided context (ctx) object and
+ * `user` and `session` attribute to the request (req) object.
+ *
+ * @internal
+ */
+export function injectAuthIntoArgs(ctx: LoaderFunctionArgs, authData: AuthData): LoaderFunctionArgsWithAuth {
+  const { user, session, userId, sessionId } = authData;
+  const auth = {
+    userId,
+    sessionId,
+    getToken: authData.getToken,
+  };
+  return { ...ctx, auth, user, session };
+}
+
+/**
+ * See `packages/nextjs/src/middleware/utils/sanitizeAuthData.ts`
+ * TODO: Make a shared function
+ *
+ * @internal
+ */
+export function sanitizeAuthData(authData: AuthData): any {
+  const user = authData.user ? { ...authData.user } : authData.user;
+  if (user) {
+    // @ts-expect-error;
+    delete user['privateMetadata'];
```

Comment:
🙃 Ideally we should control the return of privateMetadata with a query parameter from the `GET /v1/users/:id` endpoint as this can be error prone in other SSR environments.

Reply:
In theory a dev might want `privateMetadata` in the loader/action, which would be acceptable usage.  I suppose we should selectively sanitize (e.g. privateMetadata is injected when calling rootAuthLoader's callback, but it's not returned for useLoaderData to see).

Reply:
Valid point. Let's discuss offline about a more robust way to protect developers for accidental leakage of privateMetadata in the SSR generator.

Reply:
Cool, let's discuss. Main reasoning behind this is that's fairly easy for someone to mistakenly serialize the user and return it to the component.

IMO, a customer should only get access to the private metadata by opting in by passing a flag on a per loader case explicitly

Reply:
+1